server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -457,7 +457,24 @@ public NodeIndicesStats stats(CommonStatsFlags flags) {
             }
         }
 
-        return new NodeIndicesStats(commonStats, statsByShard(this, flags));
+        return new NodeIndicesStats(commonStats, statsByIndex(this, flags), statsByShard(this, flags));
+    }
+
+    static Map<Index, CommonStats> statsByIndex(final IndicesService indicesService, final CommonStatsFlags flags) {
+        final Map<Index, CommonStats> statsByIndex = new HashMap<>();
+
+        for (final IndexService indexService : indicesService) {
+            Index index = indexService.index();
+            CommonStats commonStats = CommonStats.getIndexLevelStats(indexService, flags);
+
+            if (statsByIndex.containsKey(index) == false) {
+                statsByIndex.put(index, commonStats);
```

Comment:
This whole thing could just be:
```
var existing = statsByIndex.putIfAbsent(index, commonStats);
assert existing == null;
```

We never see an index twice here do we?

Reply:
You're right, I copied it from the bit below which had shards (where there it makes sense). I will adopt your comment here, thanks!